src/opencl_common.c

```diff
@@ -2829,6 +2881,269 @@ int opencl_calc_min_kpc(size_t lws, size_t gws, int v_width)
+
+			list->count--;
+
+			/* Done! */
+			break;
+		}
+		prev = current;
+	}
+}
+
+static void krn_list_init(struct opencl_kernels **list) {
+	*list = mem_alloc(sizeof(struct opencl_kernels));
+	(*list)->tail = (*list)->head = NULL;
+	(*list)->count = 0;
+}
+
+static void krn_list_add_link(struct opencl_kernels *list, struct kernel_opencl *entry) {
+	entry->next = NULL;
+
+	if (list->tail)
+		list->tail = list->tail->next = entry;
+	else
+		list->tail = list->head = entry;
+
+	list->count++;
+}
+
+static void krn_list_add(struct opencl_kernels *list, cl_kernel *data) {
+	struct kernel_opencl *entry;
+
+	entry = mem_alloc(sizeof(struct kernel_opencl) + sizeof(data));
+	entry->data = *(data);
+	entry->variable = data;
+
+	krn_list_add_link(list, entry);
+}
+
+static void krn_list_remove(struct opencl_kernels *list, cl_kernel data) {
+	struct kernel_opencl *current, *prev = NULL;
+
+	for (current = list->head; current; current = current->next) {
+		if (current->data == data) {
+			if (prev)
+				prev->next = current->next;
+
+			if (list->head == current)
+				list->head = current->next;
+			if (list->tail == current)
+				list->tail = prev;
+
+			*(current->variable) = NULL; //Is it really safe?
+			MEM_FREE(current);
+
+			list->count--;
+
+			/* Done! */
+			break;
+		}
+		prev = current;
+	}
+}
+
+/*
+ * Use this function to alloc OpenCL buffers.
+ * Later, all buffers can be released at once using opencl_release_buf().
+ *
+ * Setting keep_track to zero (0) causes the function to NOT save a reference
+ * of the newly created buffer. Therefore, the buffer will not be released
+ * automatically by opencl_release_buf().
+ *
+ * IMPORTANT: the function keeps a referent to _buffer. So, it can't be freed
+ *             before a call to opencl_release_buf();
+ */
+cl_mem opencl_create_buf(cl_mem *_buffer, cl_mem_flags _flags, size_t _size,
+            int keep_track) {
+	cl_int __errCreateBuffer;
+
+	if (*(_buffer))
+		opencl_release_buf(_buffer);
+
+	*(_buffer) =
+		clCreateBuffer(context[gpu_id], _flags, _size, NULL, &__errCreateBuffer);
+
+	if (__errCreateBuffer != CL_SUCCESS) {
+		*(_buffer) = NULL;
+		HANDLE_CLERROR(__errCreateBuffer, "Error allocating GPU memory");
+	} else
+		if (keep_track)
+			buf_list_add(cl_buffer_list, _buffer);
+
+	return *(_buffer);
+}
+
+/*
+ * Use this function to alloc OpenCL buffers and mapped buffers.
+ * Later, all buffers can be released at once using opencl_release_map().
+ */
+void opencl_create_map(cl_mem_flags _buffer_flags, size_t _size, cl_mem *_out_buffer,
+        cl_bool _map_block, cl_map_flags _map_flags, size_t _map_offset, size_t _map_size,
+		void **_out_map) {
+
+	cl_int __errCreateBuffer, __errMapBuffer, __failure = CL_SUCCESS;
+
+	*(_out_buffer) = clCreateBuffer(context[gpu_id], _buffer_flags,
+		_size, NULL, &__errCreateBuffer);
+	*(_out_map) = clEnqueueMapBuffer(queue[gpu_id], *(_out_buffer),
+		_map_block, _map_flags, _map_offset, _map_size, 0, NULL, NULL,
+		&__errMapBuffer);
```

Comment:
Maybe I'm stupid now but isn't (for our use) always `_size` same as `_map_size`? If so we could drop the latter argument from the function.

Also, I **think** we (so far) never use `CL_MAP_READ` so perhaps we should hard-code that (perhaps even to `CL_​MAP_​WRITE_​INVALIDATE_​REGION`?) and drop the `_map_flags` argument.

Reply:
All current code has `_map_block` as `CL_TRUE` so we can optimize that away as well

Reply:
All current code has `_map_offset` as `0` so we can optimize that away as well

Reply:
Done.